granary/bluesky.py

```diff
@@ -161,25 +164,28 @@ def from_as1(obj, from_url=None):
         banner = url
         break
 
-    url = util.get_url(obj) or obj.get('id') or ''
+    url = util.get_url(obj)
+    id = obj.get('id')
+    if not url and id:
+      url = ''
+      parsed = util.parse_tag_uri(id)
+      if parsed:
+        url = f'http://{parsed[0]}'
     try:
       did_web = url_to_did_web(url)
     except ValueError as e:
       logger.info(f"Couldn't generate did:web: {e}")
       did_web = ''
 
-    # handle is username@domain or domain/path, no scheme or query
+    # handles must be hostnames
+    # https://atproto.com/specs/handle
     username = obj.get('username')
     parsed = urllib.parse.urlparse(url)
     domain = parsed.netloc
-    if username:
+    if username and HANDLE_PATTERN.search(username):
       handle = username
-      if domain:
-        handle += f'@{domain}'
     elif url:
       handle = domain
-      if parsed.path not in ('', '/'):
-        handle += parsed.path
     else:
       handle = ''
 
```

Comment:
should line 198 be `'did': id or did_web`? if the input is already an ATProto user and `id` is set, it's hopefully the right DID, right? I'm thinking of where and how we want to support `did:plc`s so that we don't assume `did:web` too often, since those are basically nonexistent in practice right now.

Reply:
Annoyingly I think `id` is a tag_uri not a did. I don't think we're going to have `did:plc`s in pure as1? We'd only really have them if we get them from Bluesky to begin with. I'd rather not try to implement their plc method algorithm!

Reply:
Yeah good point. Maybe `id if id and id.startswith('did:') else did_web`? I think you're right that our use cases right now don't include Bluesky records coming through. I guess I'm future proofing, I just don't want to blindly discard a valid DID in favor of a `did:web` we generated, esp since that won't actually resolve via the `did:web` resolution algorithm.